pkg/api/validation/events.go

```diff
@@ -25,13 +30,45 @@ import (
 // ValidateEvent makes sure that the event makes sense.
 func ValidateEvent(event *api.Event) errs.ValidationErrorList {
 	allErrs := errs.ValidationErrorList{}
-	// TODO: There is no namespace required for node.
-	if event.InvolvedObject.Kind != "Node" &&
-		event.Namespace != event.InvolvedObject.Namespace {
-		allErrs = append(allErrs, errs.NewFieldInvalid("involvedObject.namespace", event.InvolvedObject.Namespace, "namespace does not match involvedObject"))
+
+	// Todo: Events may reference 3rd party object, and we can't check whether the object is namespaced. Suppose them are namespaced. Do check if we can get the piece of information.
+	groupVersion := event.InvolvedObject.APIVersion
+	if groupVersion == "v1" || groupVersion == "extensions/v1beta1" {
+		nonNamespacedKindFlag, err := isNonNamespacedKind(event.InvolvedObject.Kind, event.InvolvedObject.APIVersion)
+		if err != nil {
+			allErrs = append(allErrs, errs.NewFieldInvalid("involvedObject.Kind", event.InvolvedObject.Kind, fmt.Sprintf("couldn't check whether namespace is allowed: %s", err)))
+		} else {
+			if !nonNamespacedKindFlag &&
+				event.Namespace != event.InvolvedObject.Namespace {
+				allErrs = append(allErrs, errs.NewFieldInvalid("involvedObject.namespace", event.InvolvedObject.Namespace, "namespace does not match involvedObject"))
+			}
```

Comment:
I think there should be a check that non-namespaced objects are having their events made in the default namespace? I think that's where we decided they go?